### ISSUE TITLE
openSUSE packages failing when built with Nuitka

### ISSUE THREAD

**Issue by jayvdb** [Nuitka/Nuitka#522]
Project https://build.opensuse.org/project/show/home:jayvdb:Nuitka - Python 3 only

The collection of packages contains the initial set of dependencies needed to bootstrap a standard Python installation, with all build dependencies and necessary testing tools needed to verify the built packages.  Still quite a few unresolved packages, preventing building of larger frameworks like Django.

Nuitka build failures:
- [ ] [testtools](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-testtools) `assert locals_dict_name not in locals_dict_handles, locals_dict_name` : `globals_testtools$_compat2x` : "`Cannot recurse to import module 'testtools/_compat2x.py' (..) because of 'SyntaxError'`" (_Fixed by deleting `testtools/_compat2x.py` on Python 3_)
- [ ] [Js2Py](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Js2Py) `assert locals_dict_name not in locals_dict_handles, locals_dict_name`: `globals_js2py$internals$speed` : "`Cannot recurse to import module 'js2py/internals/speed.py' (...) because of 'SyntaxError'`"; _bypassing that results in OOM in clang and gcc_
- [ ] [PasteDeploy](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-testtools) `assert locals_dict_name not in locals_dict_handles, locals_dict_name` : globals_paste$deploy$paster_templates (_Fixed with `mv paste/deploy/paster_templates.py paste/deploy/paster_templates/__init__.py`)
- [ ] [Pygments](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Pygments) `assert locals_dict_name not in locals_dict_handles, locals_dict_name` : `globals_tests$support` (_Fixed by removing `tests/`_)
- [ ] Twisted https://github.com/Nuitka/Nuitka/issues/531
- [ ] [datrie](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-datrie) & [numpy](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-numpy) & [greenlet](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-greenlet) `Missing both compile_packages and py_modules, aborting...`
- [ ] [pyzmq](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pyzmq) gcc: `/tmp/timer_createamusva27.c:2: undefined reference to `timer_create'`
- [ ] [PyInstaller](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-PyInstaller) `Error, 'hook-PIL.Image' is not a proper python module name.`

Missing artifacts (https://github.com/Nuitka/Nuitka/pull/530 is a workaround for someone of these):
- [ ] [Jinja2](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Jinja2) - `Jinja2.tests` not available https://github.com/Nuitka/Nuitka/issues/525
- [ ] [PySocks](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-PySocks) - `sockshandler` discarded
- [ ] [Cython](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Cython) - cython -> Cython (maybe ok?)
- [ ] [peewee](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-peewee) - packages `peewee` and `pwiz` discarded
- [ ] [cryptography](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-cryptography) - `ModuleNotFoundError: No module named 'cryptography.hazmat.bindings._constant_time'`
- [ ] [pytest-runner](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pytest-runner) ptr.so created; pytest_runner.so omitted

Seg faults
- [ ] [Werkzeug](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Werkzeug) only occurs in `test_shared_data_middleware`
- [ ] [hypothesis](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-hypothesis) - only occurs with clang, not gcc
- [ ] [Pympler](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Pympler) only occurs in test `test_stracker_create_summary` (`test_sys_mods` fails for less concerning reasons)

As I go through getting those packages to build, any which build but cause other failures I will disable "Use for Build Flag" e.g. https://build.opensuse.org/repositories/home:jayvdb:Nuitka/python-setuptools . I will attempt to adjust them so the tests fail.

Builds but breaks other packages:
- [ ] appdirs - see [path.py](https://build.opensuse.org/package/live_build_log/home:jayvdb:Nuitka/python-path.py:test/openSUSE_Tumbleweed/i586)
- [ ] [setuptools](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-setuptools)
- [ ] [py](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-py)
- [ ] [pytest](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pytest)
- [ ] [pbr](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pbr)
- [ ] [pytz](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pytz)
- [ ] [virtualenv](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-virtualenv)
- [ ] [alabaster](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-alabaster)
- [ ] [execnet](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-execnet) - `inspect.getsource(source)` fails, breaking pytest
- [ ] [testscenarios](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-)
- [ ] [kombu](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-kombu)
- [ ] [pip](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pip)
- [ ] [eventlet](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-eventlet), especially eventlet.patcher.SysModulesSaver
- [ ] `zope.*`, `sphinxcontrib.*`, `paste.*` https://github.com/Nuitka/Nuitka/issues/524
- [ ] [Werkzeug](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Werkzeug) breaks [Flask](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Flask) `AttributeError: module 'werkzeug' has no attribute 'test'`

Test failures:
- [ ] [psutil](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-psutil), 10 failures
- [ ] [xmlschema](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-xmlschema) `test_element_tree_import_script` fails
- [ ] [path.py](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-path.py) `TestSpecialPaths` case fails (5 tests)
- [ ] [Paste](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Paste) `test_egg_parser` fails
- [ ] [bs4](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-beautifulsoup4) `test_out_of_range_entity` fails
- [ ] [pytest-shutil](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pytest-shutil) `test_run_in_subprocess_passes_stderr` fails
- [ ] [Jinja2](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Jinja2) - `test_safe_format_safety` and `test_safe_format_all_okay` fail
- [ ] [kombu](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-kombu)  `test_import_no_ssl`
- [ ] [ZConfig](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-ZConfig)  `test_existing_file` & `test_existing_path`
- [ ] [Werkzeug](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-Werkzeug) `test_default_stream_factory[True-???]` (two tests) `test_shared_data_middleware` (seg fault)
- [ ] [cliff](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-cliff) lots of failures, but may be only a few common problems, and maybe VM setup related
- [ ] [scons](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/scons:test) - `test/fixture/wrapper.py` & `test/fixture/wrapper_with_args.py`
- [ ] [pytest-cov](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-pytest-cov) tests fail after `test_funcarg_not_active PASSED [ 52%]`
- [ ] [mock](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-mock) - autospec functionality: `TypeError: Nuitka doesn't support __defaults__ size changes` - c.f. [graphviz](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-graphviz) `tests/conftest.py` which uses this.

Convertion to setuptools pending:
- [ ] [PyYAML](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-PyYAML) `TypeError: dist must be a Distribution instance`
- [ ] [gpgme](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/gpgme) `error: each element of 'ext_modules' option must be an Extension instance or 2-tuple` - workaround was create alternative [python-gpg](https://build.opensuse.org/package/show/home:jayvdb:Nuitka/python-gpg) which is only the Python component, but also without the tests.

**Comment by jayvdb** [comment]
I've mostly figured out cryptography, and cryptography_vectors.

I needed to copy all of the subdirectories of `cryptography_vectors` to the top level of site-packages (hopefully not too many conflicts with other packages) -- a patch should be able to fix this easily, or a Nuitka plugin maybe.  I guess there must be an existing issue about missing data files.  https://github.com/Nuitka/Nuitka/issues/521 is an umbrella issue I created about that.

Then for `cryptography`, I first tried combinations of force-inclusions, like `--include-plugin-directory cryptography.hazmat.bindings --include-module cryptography.hazmat.bindings._padding --include-module cryptography.hazmat.bindings._constant_time --include-module cryptography.hazmat.bindings._openssl --include-module cryptography.hazmat.bindings._padding.abi3 --include-module cryptography.hazmat.bindings._constant_time.abi3 --include-module cryptography.hazmat.bindings._openssl.abi3` but that didnt work, and there wasnt any warnings about those either, which feels like it should emit something.

I also tried doing `build_ext` first, and then rename `cryptography/hazmat/bindings/_constant_time.abi3.so` ->  `cryptography/hazmat/bindings/_constant_time.so` in case the PEP 3149 names were causing the problem.  Nuitka still didnt pick up the extra .so

Then I did the following before performing the build, and success.
```sh
echo 'from cryptography.hazmat.bindings import _padding' >>  src/cryptography/hazmat/bindings/__init__.py
echo 'from cryptography.hazmat.bindings import _constant_time' >>  src/cryptography/hazmat/bindings/__init__.py
echo 'from cryptography.hazmat.bindings import _openssl' >>  src/cryptography/hazmat/bindings/__init__.py
```
